### PR TITLE
Documentation: Details/Notes regarding Rails API-only applications

### PR DESCRIPTION
**Pull request by colinross**
This is a bit of a knock on PR related to #4947 specifically addressing documentation.

I have been experimenting more and more with rails api-only application leveraging devise this PR attempts to make a solid attempt to stem the documentation issues (or really lack thereof) from creating unnecessary noise as bugs or feature requests.

### CODE REVIEW

**Comment by colinross** [comment]
I'm not quite sure how to best add this point to the documentation, but there is a slight, nuanced difference between an Rails API-only Application and an application where your resources (the protected controllers for instance) are extending from `ActionController::API`. In comes down to which controller type your devise parent controller (`ApplicationController` by default) extends from.

While outside the scope of this documentation PR, I wonder if the better way to 'address supporting' API controllers would be to issue warnings when known incompatibilities are loaded/used. e.g.
 `Rails.logger "Rememberable is not compatible with controllers extending from ActionController::API; Either disable/remove the :rememberable module in the model's configuration or extend from ActionController::Base"`

File: README.md [hunk 1]
```diff
@@ -697,7 +697,22 @@ Devise supports ActiveRecord (default) and Mongoid. To select another ORM, simpl
...
+#### Supported Authentication Strategies
+The only default configured strategy supported by devise for API-only applications is database via HTTP Auth, meaning the user is effectively authenicated on each request (a byproduct of API-only applications not supporting browser-based sessions via cookies).
+
+The devise default for HTTP Auth is disabled, so it will need to be enabled in the devise initializer for the database strategy:
+
+```ruby
+config.http_authenticatable = [:database]
+```
+
+This restriction does not limit you from implementing custom strategies, either in your application or via gem-based extensions for devise.
```
Comment (mracos) [inline README.md]: I think we can be a little bit more explicit that we are talking about `warden` strategies here, something like
```suggestion
This restriction does not limit you from implementing custom warden strategies, either in your application or via gem-based extensions for devise.
```

File: README.md [hunk 2]
```diff
@@ -697,7 +697,22 @@ Devise supports ActiveRecord (default) and Mongoid. To select another ORM, simpl
...
+A common authentication strategy for APIs is token-based authenication. For more information on extending devise to support this type of authentication and others, see the wiki article for [Simple Token Authentication Examples and alternatives](https://github.com/plataformatec/devise/wiki/How-To:-Simple-Token-Authentication-Example#alternatives).
```
Comment (mracos) [inline README.md]: Little typo 😄
But I also think that we can link [@tegon's article](http://blog.plataformatec.com.br/2019/01/custom-authentication-methods-with-devise/) about adding custom authentication strategies here. Something like:

```suggestion
A common authentication strategy for APIs is token-based authentication. For more information on extending devise to support this type of authentication and others, see the wiki article for [Simple Token Authentication Examples and alternatives](https://github.com/plataformatec/devise/wiki/How-To:-Simple-Token-Authentication-Example#alternatives) or the this blog post on [Custom authentication methods with Devise](http://blog.plataformatec.com.br/2019/01/custom-authentication-methods-with-devise/).
```

File: lib/generators/templates/README [hunk 3]
```diff
@@ -1,6 +1,6 @@
 ===============================================================================
 
-Some setup you must do manually if you haven't yet:
+Some setup is required manually if you haven't yet, depending on your application's configuration:
```
Comment (mracos) [inline lib/generators/templates/README]: This sentence looks a little bit odd to me, wdyt about something like?

```suggestion
Depending on your application's configuration some manual setup may be required:
```

**Comment by colinross** [comment]
Those edits are fine. Thanks for the time.

File: README.md [hunk 4]
```diff
@@ -697,7 +697,22 @@ Devise supports ActiveRecord (default) and Mongoid. To select another ORM, simpl
 
 ### Rails API Mode
 
-Rails 5+ has a built-in [API Mode](https://edgeguides.rubyonrails.org/api_app.html) which optimizes Rails for use as an API (only). One of the side effects is that it changes the order of the middleware stack, and this can cause problems for `Devise::Test::IntegrationHelpers`. This problem usually surfaces as an ```undefined method `[]=' for nil:NilClass``` error when using integration test helpers, such as `#sign_in`. The solution is simply to reorder the middlewares by adding the following to test.rb:
+Rails 5+ has a built-in [API Mode](https://edgeguides.rubyonrails.org/api_app.html) which optimizes Rails for use as an API (only). Devise is _somewhat_ able to handle applications that are built in this mode without additional modifications in the sense that it should not raise exceptions and the like. But some issues may still arise during `development`/`testing`, as we still don't know the full extent of this compatibility. (For more information, see #4947)
```
Comment (mracos) [inline README.md]:
```suggestion
Rails 5+ has a built-in [API Mode](https://edgeguides.rubyonrails.org/api_app.html) which optimizes Rails for use as an API (only). Devise is _somewhat_ able to handle applications that are built in this mode without additional modifications in the sense that it should not raise exceptions and the like. But some issues may still arise during `development`/`testing`, as we still don't know the full extent of this compatibility. (For more information, see [issue #4947](https://github.com/plataformatec/devise/issues/4947/))
```

File: README.md [hunk 5]
```diff
@@ -697,7 +697,22 @@ Devise supports ActiveRecord (default) and Mongoid. To select another ORM, simpl
...
+
+The devise default for HTTP Auth is disabled, so it will need to be enabled in the devise initializer for the database strategy:
+
+```ruby
+config.http_authenticatable = [:database]
+```
+
+This restriction does not limit you from implementing custom warden strategies, either in your application or via gem-based extensions for devise.
+A common authentication strategy for APIs is token-based authentication. For more information on extending devise to support this type of authentication and others, see the wiki article for [Simple Token Authentication Examples and alternatives](https://github.com/plataformatec/devise/wiki/How-To:-Simple-Token-Authentication-Example#alternatives) or the this blog post on [Custom authentication methods with Devise](http://blog.plataformatec.com.br/2019/01/custom-authentication-methods-with-devise/).
```
Comment (mracos) [inline README.md]:
```suggestion
A common authentication strategy for APIs is token-based authentication. For more information on extending devise to support this type of authentication and others, see the wiki article for [Simple Token Authentication Examples and alternatives](https://github.com/plataformatec/devise/wiki/How-To:-Simple-Token-Authentication-Example#alternatives) or this blog post on [Custom authentication methods with Devise](http://blog.plataformatec.com.br/2019/01/custom-authentication-methods-with-devise/).
```

**Comment by mracos** [comment]
Thanks for your contribution @colinross. :blue_heart:

About incompatibilities between some features of `devise` and `ActionController::API` I think a good start is to map some of them (you already mentioned `rememberable`). 

With that info, we can then decide the best course of action for each one, which could be to raise a warning as you said or even just documenting it somewhere. The point is that we would need to have this somehow mapped.

Makes sense to you?

**Comment by colinross** [comment]
That makes sense to me. I'll see what I can get started on and get a PR in the near future. I'm seeing maybe a table of 'Feature/Module' with Supported Y/N and a place to detail the caveats?  I'm pretty sure Omniauth will have a caveat or two.

Maybe a FEATURE_SUPPORT.md file? Not sure if i want to muddy up the already quite extensive README.md

**Comment by mracos** [comment]
WDYT about starting with a wiki article and then later we decide where to put it?

**Comment by colinross** [comment]
Sure.

**Comment by colinross** [comment]
Started a wiki entry

https://github.com/plataformatec/devise/wiki/API-Mode-Compatibility-Guide

**Pull request merged**